rust/typeql_lang.rs

```diff
@@ -38,70 +30,75 @@ pub mod query;
 #[macro_use]
 mod util;
 
-use common::error::ErrorMessage;
+use antlr_rust::{common_token_stream::CommonTokenStream, InputStream, Parser as ANTLRParser};
+
+pub use builder::{contains, eq, gt, gte, like, lt, lte, neq, not, rel, rule, type_, var};
+use common::Result;
 use parser::{
     error_listener::ErrorListener, syntax_error::SyntaxError, visit_eof_definables,
     visit_eof_label, visit_eof_pattern, visit_eof_patterns, visit_eof_queries, visit_eof_query,
     visit_eof_schema_rule, visit_eof_variable,
 };
-use pattern::{Label, Pattern, RuleDefinition, Variable};
+use pattern::{Definable, Label, Pattern, RuleDefinition, Variable};
 use query::Query;
+use std::{cell::RefCell, rc::Rc};
+use typeql_grammar::{typeqlrustlexer::TypeQLRustLexer, typeqlrustparser::TypeQLRustParser};
+
+fn trim_start_comments(mut s: &str) -> &str {
+    while s.starts_with('#') {
+        s = s.trim_start_matches(|c| c != '\n').trim_start();
+    }
+    s
+}
 
 macro_rules! parse {
-    ($visitor:ident($accessor:ident($input:expr))) => {{
-        let input = $input;
+    ($visitor:ident($accessor:ident($input:ident))) => {{
+        let input = trim_start_comments($input.trim());
         let lexer = TypeQLRustLexer::new(InputStream::new(input));
         let mut parser = TypeQLRustParser::new(CommonTokenStream::new(lexer));
 
         parser.remove_error_listeners();
         let errors = Rc::new(RefCell::new(Vec::<SyntaxError>::new()));
         parser.add_error_listener(Box::new(ErrorListener::new(input, errors.clone())));
 
-        let result = $visitor(parser.$accessor().unwrap());
+        let ast_root = parser.$accessor().unwrap();
 
         if errors.borrow().is_empty() {
-            result.map_err(|em| em.message)
+            $visitor(ast_root)
         } else {
```

Comment:
First check there were no syntax errors, then try to visit the root node, otherwise the tree we visit will be in an invalid state. Previously the visitors couldn't panic, so this bug was invisible.